applications/minotari_merge_mining_proxy/src/proxy.rs

```diff
@@ -296,49 +296,112 @@ impl InnerService {
-                trace!(target: LOG_TARGET, "Starting calculate achieved Tari difficultly");
+                let timer = Instant::now();
                 let diff = randomx_difficulty(
                     &tari_header,
                     &self.randomx_factory,
                     self.consensus_manager.get_genesis_block().hash(),
                     &self.consensus_manager,
                 )?;
-                info!(
-                    target: LOG_TARGET,
-                    "Difficulty achieved Tari difficultly - achieved {} vs. target {}",
-                    diff,
-                    block_data.template.tari_difficulty
-                );
+                trace!(target: LOG_TARGET, "Calculated achieved Tari difficultly in {:.2?}", timer.elapsed());
                 diff.as_u64()
             } else {
-                block_data.template.tari_difficulty
+                block_data.template.tari_target_difficulty
             };
 
             let height = tari_header_mut.height;
             info!(
                 target: LOG_TARGET,
-                "Checking if we must submit block #{} to Minotari node with achieved target {} and expected target: {}",
+                "Checking if we must submit block #{}. Difficulties - achieved target: {}, tari target: {}, pool target: {:?}",
                 height,
                 achieved_target,
-                block_data.template.tari_difficulty
+                block_data.template.tari_target_difficulty,
+                block_data.template.p2pool_target_difficulty,
             );
-            if achieved_target >= block_data.template.tari_difficulty {
-                let resp = match p2pool_client {
-                    Some(mut client) => {
-                        info!(target: LOG_TARGET, "Submiting to p2pool");
-                        client
-                            .submit_block(SubmitBlockRequest {
-                                block: Some(block_data.template.tari_block),
-
-                                wallet_payment_address: self.wallet_payment_address.to_hex(),
-                            })
-                            .await
-                    },
-                    None => base_node_client.submit_block(block_data.template.tari_block).await,
-                };
 
+            if self.p2pool_client.is_some() && block_data.template.p2pool_target_difficulty.is_none() {
+                return Err(MmProxyError::LogicalError("No node to submit block to".to_string()));
+            }
+
+            let submit_to_base_node = achieved_target >= block_data.template.tari_target_difficulty;
+            let submit_to_p2pool = p2pool_client.is_some() &&
+                achieved_target >= block_data.template.p2pool_target_difficulty.expect("has value");
+
+            let mut bn_time = Duration::from_secs(0);
+            let mut p2p_time = Duration::from_secs(0);
+            let (resp_bn, resp_p2p) = match (submit_to_base_node, submit_to_p2pool) {
```

Comment:
Dont use a match here, use two if's
You are duplicateing the submit to base and submit to p2pool code twice by using a match